oxidation/bindings/generator/templates/binding_web_meths.rs.j2

```diff
@@ -0,0 +1,395 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+{# Macros -#}
+
+{#-
+ # Types conversions macros
+ #}
+
+
+{%- macro function_param_type(type) -%}
+{%- if type.kind == "ref" -%}
+{{ function_param_type(type.elem) }}
+{%- elif type.kind == "ref" -%}
+&{{ function_param_type(type.elem) }}
+{%- elif type.kind == "list" -%}
+Vec<{{ function_param_type(type.elem) }}>
+{%- elif type.kind == "optional" -%}
+Option<{{ function_param_type(type.elem) }}>
+{%- elif type.kind == "bool" -%}
+bool
+{%- elif type.kind == "int" -%}
+i64
+{%- elif type.kind == "float" -%}
+f64
+{%- elif type.kind in ("str", "str_based") -%}
+String
+{%- elif type.kind == "bytes" -%}
+Uint8Array
+{%- elif type.kind == "struct" -%}
+Object
+{%- elif type.kind == "variant" -%}
+Object
+{%- elif type.kind == "result" -%}
+Object
+{%- else -%}
+{{ raise("Unsupported type %r" % type) }}
+{%- endif -%}
+{%- endmacro -%}
+
+
+{%- macro render_function_param_cooking(param_name, type) -%}
+{% if type.kind == "ref" %}
+{{ render_function_param_cooking(param_name, type.elem) }}
+{% elif type.kind == "optional" %}
+let {{ param_name }} = match {{ param_name }} {
+    Some({{ param_name }}) => {
+        {{ render_function_param_cooking(param_name, type.elem) | indent(8) }}
+        Some({{ param_name }})
+    }
+    None => None,
+};
```

Comment:
This doesn't work because using map creates closures that can return `X` or `Result<X, JsValue>` (depending on if X is a scalar or a more complex type that needs additional conversion)

see for instance:
```rust
#[allow(non_snake_case)]
#[wasm_bindgen]
pub fn fP10(_a: Option<bool>) -> Promise {
    future_to_promise(async move {
        let _a = _a.map(|_a| {            
            _a  // <--- Simple boolean, the map returns a bool
        });
        libparsec::f_p10(_a);
        Ok(JsValue::NULL)
    })
}

#[allow(non_snake_case)]
#[wasm_bindgen]
pub fn fP11(_a: Option<Object>) -> Promise {
    future_to_promise(async move {
        let _a = _a.map(|_a| {
            let _a = _a.into();
            let _a = struct_data_js_to_rs(_a)?;  // <--- inner type conversion, the map returns a Result<..., JsValue>
            _a
        });

        libparsec::f_p11(&_a);
        Ok(JsValue::NULL)
    })
}
```

Of course we could try to add more check to change the behavior depending of either `X` needs additional conversion or not, but this will add unneeded complexity so I'd rather stick with the match approach ;-)